### PR TITLE
changed from flask to fastapi, only problem i have left is with patch & post error

### PR DESCRIPTION
**Pull request by YuvHayun**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,17 +1,13 @@
...
-    app.run(debug=True)
 
+app = FastAPI()
 beers = {}
 
 
-@app.before_first_request
-def on_run_app():
+@app.on_event("startup")
+async def startup_event():
```
Comment (Hedingber) [inline app.py:10]: You're defining this function as `async` but you're running synchronous code inside it,
FastAPI does support sync functions (i.e. without `async`) 
but what I want is that out application will be fully async - therefore - change the code to be async

File: app.py [hunk 2]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
         beers = {}
 
 
-def on_exit_app():
+@app.on_event("shutdown")
+def shutdown_event():
```
Comment (Hedingber) [inline app.py:20]: Here you're running sync code in a sync function, but as I said, I want async code in async function

File: app.py [hunk 3]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
...
     with open("data.json", "w+") as outfile:
         json.dump(beers, outfile)
 
 
-atexit.register(on_exit_app)
+@app.get("/beers/{isbn}")
+async def get_beer(isbn: str):
+    if isbn in beers:
+        return beers[isbn]
+    raise HTTPException(status_code=404, detail="Item not found")
```
Comment (Hedingber) [inline app.py:29]: Be more "nice" to your API users, give more details in the error message
```suggestion
    raise HTTPException(status_code=404, detail=f"Beer {isbn} not found")
```

File: app.py [hunk 4]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
...
-        beer_list = filtered_beer_list
-
-    return jsonify(beer_list)
-
-
-@app.route('/beers', methods=['POST'])
-def create_beer():
-    if validate_beer_object():
+        if f'filter[{filter_name}]' in req.query_params:
+            filter_value = req.query_params[f'filter[{filter_name}]']
```
Comment (Hedingber) [inline app.py:37]: DRY - https://en.wikipedia.org/wiki/Don%27t_repeat_yourself (doing the `f'filter[{filter_name}]'` twice)
One of
```suggestion
        filter_value = req.query_params.get(f'filter[{filter_name}]')
        if filter_value is not None:
```
or
```suggestion
        filter_query_param_name = if f'filter[{filter_name}]'
        if filter_query_param_name in req.query_params:
            filter_value = req.query_params[filter_query_param_name]
```

File: app.py [hunk 5]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
...
+async def get_beer(isbn: str):
+    if isbn in beers:
+        return beers[isbn]
+    raise HTTPException(status_code=404, detail="Item not found")
 
 
-@app.route('/beers')
-def list_beer():
+@app.get("/beers")
+async def list_beer(req: Request):
```
Comment (Hedingber) [inline app.py:33]: It's list of beers (plural)
```suggestion
async def list_beers(req: Request):
```

File: app.py [hunk 6]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
...
+                if str(_beer[filter_name]) in filter_values:
+                    filtered_beer_list.append(_beer)
+            beer_list = filtered_beer_list
+    return beer_list
+
+
+@app.post("/beers")
+async def create_beer(name: str = Body(2000), brand: str = Body(2000), price: float = Body(2000), isbn: int = Body(2000)):
+    if not name or not brand or not price or not isbn:
+        raise HTTPException(status_code=400, detail="Item not found")
```
Comment (Hedingber) [inline app.py:50]: Wrong error message

File: app.py [hunk 7]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
...
+            filter_values = filter_value.split(',')
+            for _beer in beer_list:
+                if str(_beer[filter_name]) in filter_values:
+                    filtered_beer_list.append(_beer)
+            beer_list = filtered_beer_list
+    return beer_list
+
+
+@app.post("/beers")
+async def create_beer(name: str = Body(2000), brand: str = Body(2000), price: float = Body(2000), isbn: int = Body(2000)):
```
Comment (Hedingber) [inline app.py:48]: What were you trying to do with this ` = Body(2000)` ?
This is the source of your errors

File: app.py [hunk 8]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
...
-        return True
-    return False
+        beers[str(isbn)] = new_beer
+        raise HTTPException(status_code=204)
 
 
-@app.route('/beers/<string:isbn>')
-def get_beer(isbn):
+@app.patch("/beers/{isbn}")
+async def update_beer(isbn: str, name: str = Body(''), brand: str = Body(''), price: float = Body('')):
```
Comment (Hedingber) [inline app.py:63]: what is that `= Body('')` suppost to do ?

File: app.py [hunk 9]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
...
-@app.route('/beers/<string:isbn>', methods=['DELETE'])
-def delete_beer(isbn):
+        if name:
+            beers[isbn]["name"] = name
+        if brand:
+            beers[isbn]["brand"] = brand
+        if name:
+            beers[isbn]["price"] = price
+        raise HTTPException(status_code=204)
+    raise HTTPException(status_code=400, detail="Item not found")
```
Comment (Hedingber) [inline app.py:72]: Wrong status code, extend message

File: app.py [hunk 10]
```diff
@@ -20,87 +16,66 @@ def on_run_app():
...
+
+@app.delete("/beers/{isbn}")
+async def delete_beer(isbn: str):
     if isbn in beers:
         del beers[isbn]
-        return Response("", status=204)
-    return Response('', status=404)
-
+        raise HTTPException(status_code=204)
+    raise HTTPException(status_code=404, detail="Item not found")
```
Comment (Hedingber) [inline app.py:80]: There are 2 approaches here:
1. Explicit - If someone asks to delete X but X is not found I return error code (cause I didn't delete)
2. Implicit - If someone asks to delete X but X is not found I return success (cause bottomline X doesn't exist)
both are fine, though a lot of times it's more comfortable to go with 2
choose what you want

File: requirements.txt [hunk 11]
```diff
@@ -1 +1,2 @@
-Flask==1.1.2
+fastapi==0.61.1
+uvicorn==0.12.1
```
Comment (Hedingber) [inline requirements.txt:2]: You've added uvicorn, so I guess you're using it to run the app, but I don't see that you've updated the README.md with instructions on how to run it. (instructing to do something like `uvicorn app.main:app --reload`(
There is a way to configure uvicorn to be the runner from code, so you can still run the app by doing `python app.py`, please find how, and do it

**Review by Hedingber — CHANGES_REQUESTED** [review]

File: app.py (hunk 7 above)
Comment (YuvHayun) [reply app.py:48]: i tried several options.. this is just the version i uploaded, i also tried gt=0 adn more

File: app.py (hunk 10 above)
Comment (YuvHayun) [reply app.py:80]: i'll stay with the explicit approach for now

**Pull request closed without merging**